src/cluster.py

```diff
@@ -0,0 +1,226 @@
+    """Exception raised when there is already a running cluster."""
+
+    pass
+
+
+class ClusterCreateError(Exception):
+    """Exception raised by an error on cluster creation."""
+
+    pass
+
+
+class ClusterNotRunningError(Exception):
+    """Exception raised when cluster is not running after start call."""
+
+    pass
+
+
+class ClusterStartError(Exception):
+    """Exception raised by an error on cluster start."""
+
+    pass
+
+
+class PostgresqlCluster:
+    """This class handles the creation, start and listing of PostgreSQL clusters.
+
+    A PostgreSQL cluster is a collection of databases that is managed by a single instance of a
+    running database server.
+    """
+
+    def __init__(self, version):
+        self.version = version
+        self.conf_path = Path(f"/etc/postgresql/{version}/main")
+
+    def bootstrap_cluster(self, password: str) -> None:
+        """Bootstrap a PostgreSQL cluster with the given superuser password."""
+        # Check for no running clusters (like the default cluster created on postgres install).
+        if self._is_cluster_running():
+            raise ClusterAlreadyRunningError()
+        else:
+            # Create a new cluster.
+            self._create_cluster(password)
+            # Render the configuration files and start the cluster.
+            self._render_pg_hba_conf_file()
+            self._render_postgresql_conf_file()
+            self._start_cluster()
+            # Check that the cluster is up and running.
+            if not self._is_cluster_running():
+                raise ClusterNotRunningError()
+
+    def inhibit_default_cluster_creation(self) -> None:
+        """Stop the PostgreSQL packages from creating the default cluster."""
+        os.makedirs(os.path.dirname(CREATE_CLUSTER_CONF_PATH), mode=0o755, exist_ok=True)
+        with open(CREATE_CLUSTER_CONF_PATH, mode="w") as file:
+            file.write("create_main_cluster = false\n")
+            file.write(f"include '{self.conf_path}/conf.d/postgresql-operator.conf'")
+
+    def _create_cluster(self, password: str) -> None:
+        """Create a new PostgreSQL cluster."""
+        # Write the password in a temporary file for pg_createcluster.
+        with open("/tmp/pwfile", mode="w") as file:
+            file.write(password)
+        try:
+            command = [
+                "pg_createcluster",
+                self.version,
+                "main",
+                "--datadir=/var/lib/postgresql/data/pgdata",
+                "--",
+                "--pwfile=/tmp/pwfile",
+            ]
+            logger.info(f"pg_createcluster call: {' '.join(command)}")
+            subprocess.check_call(command)
+        except subprocess.CalledProcessError as e:
+            raise ClusterCreateError(e.stdout)
+        finally:
+            # Remove the temporary file.
+            os.remove("/tmp/pwfile")
+
+    def _is_cluster_running(self) -> bool:
+        """Check whether the cluster is running or not."""
```

Comment:
Just an FYI - not sure if you are aware of [py-postgresql](https://pypi.org/project/py-postgresql/). While it also has database drivers to query and talk with the database, it has some built-ins regarding [cluster management](https://pythonhosted.org/py-postgresql/cluster.html) which may be more convenient; however I think it does raw process management rather than leveraging systemd (though I'm sure that could be subclassed and extended that way, and even possibly contributed back upstream).

The advantage of using this library is that there's already users who are using it and helping to work out some various bugs on connections, cluster status, etc, etc.

Reply:
Great Billy. I agree with what you wrote and it's really good to know about this package. I tested it but found some problems for our use case. For the first one, related to how Postgres version is displayed on Ubuntu, I submitted a [PR](https://github.com/python-postgres/fe/pull/116), which needs the approval of the repo maintainers. The second problem is related to run the library code as postgres user instead of root, which can be achieved by using su like you said in our meeting. As we talked, we can address this change in a future PR.

Reply:
That's unfortunate, but yeah - its not something required here anyways. I was beginning to wonder about k8s usage as well.